lib/ide/src/controller/text.rs

```diff
@@ -0,0 +1,150 @@
+//! Text Controller module.
+//!
+//! Facade over filesystem API or module text API for text editor. Does discerning between Luna
+//! module file and plain text file. In case of luna module idmap and metadata are hidden for the
+//! user.
+
+use crate::prelude::*;
+
+use crate::controller;
+
+use failure::_core::fmt::{Formatter, Error};
+use flo_stream::{Publisher, Subscriber};
+use flo_stream::MessagePublisher;
+use file_manager_client as fmc;
+use json_rpc::error::RpcError;
+use shapely::shared;
+
+
+
+// ====================
+// === Notification ===
+// ====================
+
+/// A buffer size for notification publisher.
+///
+/// We don't expect much traffic on file notifications, therefore there is no need for setting big
+/// buffers.
+const NOTIFICATION_BUFFER_SIZE : usize = 36;
+
+/// A notification from TextController.
+#[derive(Clone,Debug)]
+pub enum Notification {
+    /// File contents needs to be set to the following due to synchronization with external state.
+    SetNewContent(String),
+}
+
+
+
+// =======================
+// === Text Controller ===
+// =======================
+
+/// A handle for file.
+///
+/// This makes distinction between module and plain text files. The module files are handled by
+/// Module Controller, the plain text files are handled directly by File Manager Client.
+#[derive(Clone,Debug)]
+enum FileHandle {
+    PlainText {path:fmc::Path, file_manager:fmc::Handle},
+    Module    {controller:controller::module::Handle},
+}
+
+
+shared! { Handle
+
+    /// Data stored by the text controller.
+    pub struct State {
+        file: FileHandle,
+        /// Sink where we put events to be consumed by the view.
+        notification_publisher: Publisher<Notification>,
+    }
+
+    impl {
+        /// Get subscriber receiving controller's notifications.
+        pub fn subscribe(&mut self) -> Subscriber<Notification> {
+            self.notification_publisher.subscribe()
+        }
+
+        /// Get clone of file path handled by this controller.
+        pub fn file_path_clone(&self) -> fmc::Path {
```

Comment:
the same as above, I would name it `file_path` (no suffix `_clone`), as it suggest there is other way of accessing it. If we are concerned about the performance, lets use `Rc` inside of `Path` to make the clone super-fast.